[WIP] DIP2/DIP3/DIP4 Implementation

codablock opened the pull request:
# Introduction
This is my current version of the DIP2/DIP3/DIP4 implementation. It is not meant to be merged for 12.3 but for a later version (most likely 12.4)
Links to DIPs:
[DIP2 — Special Transactions](https://github.com/dashpay/dips/blob/master/dip-0002.md)
[DIP3 — Deterministic Masternode Lists](https://github.com/dashpay/dips/blob/master/dip-0003.md)
[DIP4 — Simplified Verification of Deterministic Masternode Lists](https://github.com/dashpay/dips/blob/master/dip-0004.md)

# Note for previous and new reviewers
Originally, this code was part of a PR in our private dashevo repo. It is however cleaned up (squashed and rebased all the review related commits) and also contains the DIP4 implementation, which was missing in the initial PR. It also contains a few more fixes and a refactoring to optimize the performance of deterministic MN bookkeeping. I decided to not squash these new fixes and refactorings into the older commits to make life easier for reviewers who already reviewed the old version of the PR. For these reviewers, the new changes start from commit 2ffdb238676d462683739a075bfc27fa06cde8d6 ("Implement and enforce CbTx with correct block height and deprecate BIP34").

# Deployment
These DIPs will be deployed through a combination of a BIP9 style deployment and a spork (SPORK_15_DETERMINISTIC_MNS_ENABLED). When the BIP9 deployment is activated, MNOs will be allowe to broadcast DIP3 special transactions and miners will be allowed to mine DIP2/DIP3 special transactions. This will also activate the bookkeeping (register, update, remove, ...) of deterministic masternodes.

It will however NOT start the use of deterministic masternodes and the old type of MNB/MNP based masternode system will keep being active. Only when spork15 gets activated, the whole network will switch over to use deterministic masternodes for all kinds of MN operations.

Between the BIP9 activation and the spork15 activation, MNOs are expected to gradually move their MNs to be ProTx based. They'd still have to issue "mn-start" on their masternodes as the old system is still active at this stage. A ProTx is internally just a normal transaction with a special payload, and thus is still a valid collateral for a MN of the old system. Between the BIP9 activation and spork15 activation, masternode broadcasts will be rejected if the collateral was created after BIP9 activation and is not a valid ProTx.

The reason for this multi stage deployment is that we must ensure that enough MNOs have switched to deterministic masternodes first before we actually switch over to use the new deterministic list. This should ensure a smooth migration without interruptions.

Spork15 is block height based and not time based (as other sporks usually are). When we see enough MNs being upgraded to ProTX, we will set spork15 to a block height in the near future. When that height is reached, all nodes in the network will drop the legacy masternode list and switch to the deterministic list. At the same time, all communication in regard to MNBs, MNWs and MNPs will be stopped. Also, mnsync will slightly change so that it completely skips the MASTERNODE_SYNC_LIST and MASTERNODE_SYNC_MNW state.

The masternode.conf file will then be ignored and masternode start-xxx won't be required anymore. As long as the MNO has -masternodeprivkey=<key> correctly set, the node will immediately start functioning when it encounters the ProTX on-chain. If the node was started correctly and functioning as a legacy MN at the time of spork15 activation, it will switch to the deterministic mode automatically without the MNOs intervention.

# Code-level compatibility between legacy and deterministic mode
As DIP3 has effects on many parts of the code and we at the same time need to support legacy and deterministic mode in the same code base, I decided to not touch too much of the legacy code and design. Instead, I tried to keep the interfaces of classes like CMasternode, CMasternodeMan, CMasternodePayments, ... as compatible as possible with the legacy code and implement the switching logic into these classes. The result is that these classes now contain some ugly if/else statements but at the same time all the other code that uses these classes is mostly untouched. IMHO, this reduces the risk of introduced bugs dramatically. After spork15 activation, my next task will be to clean up the code again and remove all the legacy code so that we have a clear codebase again.

# Maintaining the deterministic MN list
Each block might result in a new version of the MN list. These lists are maintained by CDeterministicMNListManager. ProcessBlock is called for each block, which loops through all ProTx of the block and creates a new MN list based on the previous list and the freshly gathered information. It also scans for spends of ProRegTx collaterals and removes MNs if needed. The information about what has been added/updated/removed is stored in leveldb once per block (with CDeterministicMNListDiff). 

Every 576 blocks, a snapshot of the list of the current block is stored in leveldb. Based on the snapshots and the diffs that follow, the list for every block height can be reconstructed (see GetListForBlock). A cache is used to avoid doing this too often.

A list is represented by CDeterministicMNList. It internally uses pure functional maps (immer::map) and shared pointers to hold the actual list and information about each entry. (Note for previous reviewers: this was different in the initial version). The reason for these data structures is that it simplifies the deduplication of slightly changed lists between blocks, which in turn dramatically reduces the memory usage.

codablock commented:
FYI: I'm currently working on fixing Travis builds, which are failing due to the use of C++14 features and Travis using Ubuntu Trusty. Please ignore the WIP commits at the moment. I'll clean them up later and put them into a separate PR.

codablock commented:
> This is a shame. You are trying to write the IP addresses of the masternodes in the blockchain!
This is an awfull crime against privacy.
THIS COIN USED TO BE DARKCOIN ONCE UPON TIMES!!!
How it ended like this??
Even Vitalik is a pro-privacy now, but the dashcore team goes against the stream!
https://twitter.com/VitalikButerin/status/1000869998385074176

Disclaimer: demodun is s known troll in the Dash forums and he/she tends to spam and derail whole threads with dozens of posts containing unrelated references to other topics and Youtube videos. Insults are also one of his strength. I expect him to do the same here after his initial post. If he does this as expected, I'll delete all his comments except the first one. @demodun If you want to go this way again, move the convo to the forums.

Now the actual answer to your concerns. Yes this will write IPs to the chain, the reason is to have a verifiable masternode list for all clients, including SPV clients. But this is about masternodes, which are expected to be publicly known and serving under a publicly known IP which can be related to a publicly known collateral. This was already the case before these DIPs and they don't change this fact. The difference is, that previously IPs were broadcasted TOGETHER with the information about the collateral and now they are stored on chain. This means: The IPs were always trackable, someone who is interested just had to connect a normal node and regularly dump the result of `masternode list`. This all applies to masternodes only, normal and SPV nodes are not affected and they will keep their privacy.

The privacy in Dash is mainly for normal users, privacy of masternodes is not as important IMHO. And as already said, these DIPs do not introduce this, this was always the case. If you run a masternode and want to get your rewards, you are expected to be publicly reachable and verifiable. Otherwise, how should normal or SPV nodes figure out to which masternodes they should connect?

demodun commented:
I refuse to answer to you false accusation about trollling, unless you support your claims with urls where I am supposed to troll.

Privacy for Masternodes IS important. This is what masternodes have voted.
https://www.dash.org/forum/threads/poll-mn-operators-please-respond.8688/

But **the core team ignored the decision of the masternodes** 
And now, by writing the IPs the blockchain, you create a permanent file of UNDENIABLE IP tracking.
The privacy of masrternodes is DEAD that way, and of course also the privacy of their vote is ALSO dead.
This is a shame for the core team to expose the IPs of the masternodes in such a way and tottaly against their will.
Shame on you, and shame to all the core team.

UdjinM6 commented:
This PR is for an _implementation_ of DIP4. If you have some questions/concerns regarding DIP4 itself, you should create an issue at https://github.com/dashpay/dips/issues or start corresponding discussion on some forum/chat/etc. Please note that this is the first version of DIPs, so if you have some suggestions how to improve DIPs while preserving the desired functionality, pls share you ideas on one of the media I pointed out above.

Having said this, pls keep this PR implementation focused. Thanks.

demodun commented:
@UdjinM6 and all the others who agree with his suggestion to put my concerns in  https://github.com/dashpay/dips/issues I want to ask you a single question:
**Do you expect for me to help improving dash, while at the same time you have ban my 2_years_old/3000_messages account from www.dashtalk.org without even give a public reason for that????**

demodun commented:
@schinzelh if this is not the place to discuss my ban from dashtalk, where can I discuss it???
**The ban occurred after I complained about @codablock who exposes the IP addresses of the masternodes in a permanent undeniable record file into the blockchain.**

So please suggest me a place to discuss the ban problem, and do not suggest me to send an e-mail in dashtalk admin, as long as they have banned me WITHOUT GIVING A PUBLIC REASON.

PaulieD commented:
@demodun... Hmm... I don't know. Maybe on the [forum](forum.dash.org), which by the way means "a public meeting place for open discussion." (https://www.merriam-webster.com/dictionary/forum) Seems to me like that fits what you're looking for, instead of flooding the github and notifying users such as myself and many others who are trying to learn more about the tech, not about your oh so terrible ban of a "**2_years_old/3000_messages**" account (wtf is that formatting anyways? underscores and bolded... so trashy). 

So yeah, in recap. Please kindly remove yourself from talking about your petty politics in an implementation focused communication channel and instead talk about the plethora of issues in your life over on the forum. Thanks. 

On a more positive note, @codablock this works looks great and I am glad to be able to be here and look through it all. Keep doing what you're doing! I love it!

demodun2 commented:
@PaulieD please try to understand that the dash forum, the dash discord and the dashpay in reddit are highly conrolled by the same core team who practices severe censorship.
So I can only talk to neurtal places, like this one or in bitcointalk.
If you really care to understand the underlying technology, you have to liste to everyone and not only the dash core team propaganda

sidhujag commented:
hmm never really thought about it but yea makes sense that before the ip info would be erased through history giving plausible deniability whilst on chain now removes that so if masternodes are caught trafficking bad data im whichever context they may be liable legally.. far fetched but never know. Maybe the ip data can still live offchain.. anchored to your onchain data store through hash. Consensus on ip is not mandatory but verifiability involves mostly just the address and public key and not auxiliary masternode data as in this case unless you have a usecase where ip is needed for historical verifiability in some context?

nmarley commented:
@sidhujag Thank you for your input! Not sure if you've read https://github.com/dashpay/dips/blob/master/dip-0003.md, but this is actually the implementation of that (and DIPs 2 and 4 as well, all at once).

If we could keep design discussion over there at the DIPs repository (feel free to open an issue), it would help us keep this particular discussion focused on the code, and specifically the code changes in this PR. Really appreciate your input and help! Thanks! 😄

codablock commented on qa/rpc-tests/test_framework/mininode.py:
```diff
@@ -357,33 +357,44 @@ class CTransaction(object):
 
     def deserialize(self, f):
-        self.nVersion = struct.unpack("<i", f.read(4))[0]
+        ver32bit = struct.unpack("<i", f.read(4))[0]
```
You're right, but I found this more explicit in what is happening there.

codablock commented on qa/rpc-tests/test_framework/mininode.py:
```diff
@@ -357,33 +357,44 @@ class CTransaction(object):
-        self.nVersion = struct.unpack("<i", f.read(4))[0]
+        ver32bit = struct.unpack("<i", f.read(4))[0]
+        self.nVersion = ver32bit & 0xffff
+        self.nType = (ver32bit >> 16) & 0xffff
```
I've got burned by C/C++ in the past when it comes to right shifting of (possibly) negative numbers as it extends the sign bit. This means, 1110 >> 1 = 1111 and not 0111 as someone might expect. And as I understand [this](https://stackoverflow.com/questions/26521412/python-bit-shifting-negative-vs-positive-values), it's the same in python.

demodun3 commented:
So the dash core team reported my account , and it was deleted!
Thas funny!
They are afraid so much my words, and my reports on how the masternodes are voting, dont they?
Stupids! Delete my account as many times as you want.
The truth cannot be hided, and it will burn you!

demodun3 commented:
So you deleted my comment, but I can add another one, cant I?

demodun3 commented:
And you can delete my demodun3, but I can add a demodun4 cant?

demodun4 commented:
This is a public repository. If you dont like people to comment on your work, why dont you pay 7$/month and hide in a private repository?

demodun5 commented:
I am waiting for your explanation why the demo@dashtalk account (having 3000 comments in 2 years) and the demodun@github account (that was hosting mnowatch and the tezos wiki) were reported by you and tottaly deleted.
I will keep asking about it, until I get an answer.

demodun5 commented:
Especially I want someone to take the responsibility for the deletion. Who is he?
Is it you @schinzelh ?

cronos-polis commented:
@codablock 
So, it seems you are taking Masternode List to a consensus level. 
That's actually what you suggested in this #1894 PR. 
Seems that a Masternode Limit will be easier this time. 
Great for the Dash Core Team.
I just have one question. Is this a progress to enable Collateralized Mining? Or that will be through another DIP?

PaulieD commented on src/activemasternode.cpp:
```diff
@@ -8,29 +8,158 @@
+        return;
+
+    if (!GetLocalAddress(activeMasternode.service)) {
+        state = MASTERNODE_ERROR;
```
Souldn't `strError` be set to something here?

codablock commented:
@UdjinM6 Thanks for the review and fixes :+1:, going through them now :)

Regarding `immer`. The reason I introduced it was the high use of memory when the MN list gets too long. Every time something is changed in any DMN state, the whole DMN map has to be copied for the new block. That means that for every block, we get an overhead of 32+8+8=48 bytes (hash key + smart pointer + map overhead) per entry. So with 4700 MNs, that's around 225k per map. I'd like to keep a day worth of MN lists in memory to ensure we never have to rebuild MN lists on-demand, which means we'd already need 126M just for maintaining the MN lists. This doesn't sound much, but it kills everything when integration tests with multiple nodes or a devnet with thousands of nodes is run.

I initially assumed that the DMN map only had to be copied when new MNs are added/removed/banned, so I used std::map in the beginning. I then realized that the DMN map actually has to be copied every single block due to reward payments modifying the DMN state. I then tried to optimize this by changing how stuff is stored, but then realized that whatever I came up with unnecessarily complicates things. I've been a big fan of immutable functional data structures since quite some time already (thanks to Scala), as these data structures remove a lot of the burden in such use cases. So I found `immer`, which offers this for C++. So, with this library, I'm creating an immutable "copy" of the map for every block, without actually copying it. Internally, the new map is just a reference to the old map with information about what has changed, so it just looks like a copy of the original map. `immer` will internally do all the memory management, reference counting, garbage collection, optimization and so on to keep a balance between performance and memory overhead.

I believe that the introduction of immutable functional data structures, a few other places could also be optimized. Can't name real example now however as this is more like a gut feeling ;)

UdjinM6 commented:
That's interesting, thanks for explaining! 👍

codablock commented:
Rebased on develop and added a few fixes that were required after that. The're also a few WIPs on-top, please ignore them for now as I'm going to clean them up later.

codablock commented:
@UdjinM6 Going through the other stuff you mentioned atm. I applied https://github.com/UdjinM6/dash/commit/b8c0f425bbcecb95c0ee2df1345298364faef163 locally and did a speed comparison. With forced/skipping of masternode sync, it's 277 seconds to run all tests, with proper syncing it's 384 seconds. What do you think? In the end, both ways are not so realistic compared to testnet or mainnet as they both have some "if testnet then do something else" logic.

UdjinM6 commented:
re regtes sync: the difference is that with https://github.com/UdjinM6/dash/commit/b8c0f425bbcecb95c0ee2df1345298364faef163 it actually does smth 😛`force_finish_*` just skips all the logic and no messages are sent.

UdjinM6 commented:
Thinking about it a bit more - alternatively, we could change the way `mnsync next` works to speed things up:

```diff
diff --git a/src/rpc/misc.cpp b/src/rpc/misc.cpp
index 7c6f735f2..e0f967e05 100644
--- a/src/rpc/misc.cpp
+++ b/src/rpc/misc.cpp
@@ -167,6 +167,7 @@ UniValue mnsync(const JSONRPCRequest& request)
     if(strMode == "next")
     {
         masternodeSync.SwitchToNextAsset(*g_connman);
+        masternodeSync.ProcessTick(*g_connman);
         return "sync updated to " + masternodeSync.GetAssetName();
     }
 
```

With this ^^^ `force_finish_*` should do the job too I guess.

codablock commented:
@UdjinM6 applied your diff in regard to mnsync. Also rebased on develop.

UdjinM6 commented:
A couple of patches as a result of local debugging:
- https://github.com/UdjinM6/dash/commit/d82885dab02e95cb82ba3cfddb41e77913cc6bbb (Fix sync, see #2241)
- https://github.com/UdjinM6/dash/commit/0b7daff94b1fb85efeef43718764014bfea6ba90 (Rewrite `CDeterministicMNManager::GetListForBlock` in a non-recursive way)
- https://github.com/UdjinM6/dash/commit/e4ef4ee539145563f3f9ae48f515a73ff7162a6d (Add sleep intervals, lower cpu usage + helps finding issues by making logs much less spammy in debug mode)

codablock commented:
Closing as DIP2, DIP3 and DIP4 have been merged into develop as part of separate PRs.
